pkg/repository/prisma/buffered.go

```diff
@@ -0,0 +1,276 @@
+}
+
+func (b *IngestBuf[T, U]) safeIncSizeOfData(size int) {
+	b.lock.Lock()
+	b.sizeOfData += size
+	b.lock.Unlock()
+}
+
+func (b *IngestBuf[T, U]) safeDecSizeOfData(size int) {
+	b.lock.Lock()
+	b.sizeOfData -= size
+	b.lock.Unlock()
+}
+
+func (b *IngestBuf[T, U]) safeSetLastFlush(t time.Time) {
+	b.lock.Lock()
+	b.lastFlush = t
+	b.lock.Unlock()
+}
+
+func (b *IngestBuf[T, U]) safeFetchLastFlush() time.Time {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+	return b.lastFlush
+}
+
+func (b *IngestBuf[T, U]) buffWorker() {
+	for {
+		select {
+		case <-b.ctx.Done():
+			return
+		case e := <-b.inputChan:
+			b.internalArr = append(b.internalArr, e)
+			b.safeIncSizeOfData(b.calcSizeOfData([]T{e.item}))
+
+			if len(b.internalArr) >= b.maxCapacity {
+				go b.flush(b.sliceInternalArray())
+			}
+			if b.safeFetchSizeOfData() >= b.maxDataSizeInQueue {
+				go b.flush(b.sliceInternalArray())
+			}
+
+		case <-time.After(time.Until(b.safeFetchLastFlush().Add(b.flushPeriod))):
+			if len(b.internalArr) > 0 {
+				go b.flush(b.sliceInternalArray())
+			} else {
+				b.safeSetLastFlush(time.Now())
+			}
+		}
+	}
+}
+
+func (b *IngestBuf[T, U]) sliceInternalArray() (items []*inputWrapper[T, U]) {
+	b.lock.Lock()
+	defer b.lock.Unlock()
+
+	if len(b.internalArr) >= b.maxCapacity {
+		items = b.internalArr[:b.maxCapacity]
+		b.internalArr = b.internalArr[b.maxCapacity:]
+	} else {
+		items = b.internalArr
+		b.internalArr = nil
+	}
+	return items
+}
+
+type flushResponse[U any] struct {
+	result U
+	err    error
+}
+
+func (b *IngestBuf[T, U]) calcSizeOfData(items []T) int {
+	size := 0
+	for _, item := range items {
+		size += b.sizeFunc(item)
+	}
+	return size
+}
+
+func (b *IngestBuf[T, U]) flush(items []*inputWrapper[T, U]) {
+	numItems := len(items)
+	b.safeSetLastFlush(time.Now())
+
+	var doneChans []chan<- *flushResponse[U]
+	opts := make([]T, numItems)
+
+	for i := 0; i < numItems; i++ {
+		opts[i] = items[i].item
+		doneChans = append(doneChans, items[i].doneChan)
+	}
+
+	b.safeDecSizeOfData(b.calcSizeOfData(opts))
+
+	ctx := context.Background()
+	result, err := b.outputFunc(ctx, opts)
+
+	if err != nil {
+		for _, doneChan := range doneChans {
+			select {
+			case doneChan <- &flushResponse[U]{err: err}:
+			default:
+				b.l.Error().Msgf("could not send error to done chan: %v", err)
+
+			}
+		}
+		return
+	}
+
+	for i, d := range doneChans {
+		select {
+		case d <- &flushResponse[U]{result: result[i], err: nil}:
+		default:
+			b.l.Error().Msg("could not send done to done chan")
+		}
+	}
+
+	b.l.Debug().Msgf("Flushed %d items", numItems)
+}
+
+func (b *IngestBuf[T, U]) cleanup() error {
+	g := errgroup.Group{}
+
+	for len(b.internalArr) > 0 {
+		g.Go(func() error {
+			b.flush(b.sliceInternalArray())
+			return nil
+		})
+	}
+
+	if err := g.Wait(); err != nil {
+		return err
+	}
+	b.cancel()
+
+	return nil
+}
+
+func (b *IngestBuf[T, U]) Start() (func() error, error) {
+	b.l.Debug().Msg("Starting buffer")
+
+	b.lock.Lock()
```

Comment:
nit: `defer b.lock.Unlock()` instead of manually releasing the locks to catch both cases

Reply:
done